rethinkdb/tests/cluster.py

```diff
@@ -0,0 +1,129 @@
+
+        response = conn.run(r.db(DATABASE).table(HEROES_TABLE).wait(timeout=1))
+        assert response['ready'] == 1
+
+        # Users.
+        # See: https://rethinkdb.com/docs/permissions-and-accounts/
+        response = conn.run(r.db('rethinkdb').table('users').insert({'id': AGENT_USER, 'password': AGENT_PASSWORD}))
+        assert response['inserted'] == 1
+        response = conn.run(r.db('rethinkdb').grant(AGENT_USER, {'read': True}))
+        assert response['granted'] == 1
+
+        response = conn.run(r.db('rethinkdb').table('users').insert({'id': CLIENT_USER, 'password': False}))
+        assert response['inserted'] == 1
+        response = conn.run(r.db(DATABASE).grant(CLIENT_USER, {'read': True, 'write': True}))
+        assert response['granted'] == 1
+
+    # Simulate client activity.
+    # NOTE: ensures that 'written_docs_*' and 'read_docs_*' metrics have non-zero values.
+
+    with RethinkDBConnection(r.connect(host=HOST, port=SERVER_PORTS['proxy'], user=CLIENT_USER)) as conn:
+        response = conn.run(r.db(DATABASE).table(HEROES_TABLE).insert(HEROES_TABLE_DOCUMENTS))
+        assert response['inserted'] == len(HEROES_TABLE_DOCUMENTS)
+
+        documents = list(conn.run(r.db(DATABASE).table(HEROES_TABLE)))
+        assert len(documents) == len(HEROES_TABLE_DOCUMENTS)
+
+
+@contextmanager
+def temporarily_disconnect_server(server):
+    # type: (str) -> Iterator[None]
+    """
+    Gracefully disconnect a server from the cluster.
+    Ensures that the stable is left in a stable state inside and after exiting the context.
+    """
+    service = 'rethinkdb-{}'.format(server)
+    logger.debug('temporarily_disconnect_server server=%r service=%r', server, service)
+
+    def _server_exists(conn):
+        # type: (Connection) -> bool
+        servers = conn.run(r.db('rethinkdb').table('server_status').map(r.row['name']))  # type: List[str]
+        logger.debug('server_exists server=%r servers=%r', server, servers)
+        return server in servers
+
+    def _leader_election_done(conn):
+        # type: (Connection) -> bool
+        STABLE_REPLICA_STATES = {'ready', 'waiting_for_primary', 'disconnected'}
+
+        replica_states = list(
+            conn.run(
+                r.db('rethinkdb')
+                .table('table_status')
+                .concat_map(r.row['shards'])
+                .concat_map(r.row['replicas'])
+                .map(r.row['state'])
+            )
+        )  # type: List[str]
+
+        logger.debug('replica_states %r', replica_states)
+
+        return all(state in STABLE_REPLICA_STATES for state in replica_states)
+
+    def _server_disconnected(conn):
+        # type: (Connection) -> bool
+        return not _server_exists(conn) and _leader_election_done(conn)
+
+    def _server_reconnected(conn):
+        # type: (Connection) -> bool
+        return _server_exists(conn) and _leader_election_done(conn)
+
+    with temporarily_stop_service(service, compose_file=COMPOSE_FILE):
```

Comment:
What's the purpose?

Reply:
This is used to test the behavior of the check when one of the servers in the cluster goes down, see [`test_check_with_disconnected_server`](https://github.com/DataDog/integrations-core/pull/5715/files#diff-982dcb0e148d94ed5ee681c16feebfdaR111).